File: packages/sequent-core/src/types/communications.rs
// SPDX-FileCopyrightText: 2023 Eduardo Robles <edu@sequentech.io>
// SPDX-FileCopyrightText: 2023 Felix Robles <felix@sequentech.io>
//
// SPDX-License-Identifier: AGPL-3.0-only
use serde::{Deserialize, Serialize};
use strum_macros::{Display, EnumString};

#[allow(non_camel_case_types)]
#[derive(
    Display, Serialize, Deserialize, Debug, PartialEq, Eq, Clone, EnumString,
)]
pub enum AudienceSelection {
    #[strum(serialize = "ALL_USERS")]
    ALL_USERS,
    #[strum(serialize = "NOT_VOTED")]
    NOT_VOTED,
    #[strum(serialize = "VOTED")]
    VOTED,
    #[strum(serialize = "SELECTED")]
    SELECTED,
}

#[allow(non_camel_case_types)]
#[derive(
    Display, Serialize, Deserialize, Debug, PartialEq, Eq, Clone, EnumString,
)]
pub enum CommunicationType {
    #[strum(serialize = "CREDENTIALS")]
    CREDENTIALS,
    #[strum(serialize = "BALLOT_RECEIPT")]
    BALLOT_RECEIPT,
    #[strum(serialize = "PARTICIPATION_REPORT")]
    PARTICIPATION_REPORT,
    #[strum(serialize = "ELECTORAL_RESULTS")]
    ELECTORAL_RESULTS,
    #[strum(serialize = "OTP")]
    OTP,
    #[strum(serialize = "TALLY_REPORT")]
    TALLY_REPORT,
}

#[allow(non_camel_case_types)]
#[derive(
    Display, Serialize, Deserialize, Debug, PartialEq, Eq, Clone, EnumString,
)]
pub enum CommunicationMethod {
    #[strum(serialize = "EMAIL")]
    EMAIL,
    #[strum(serialize = "SMS")]
    SMS,
    #[strum(serialize = "DOCUMENT")]
    DOCUMENT,
}

#[derive(Deserialize, Debug, Serialize, Clone)]
pub struct EmailConfig {
    pub subject: String,
    pub plaintext_body: String,
    pub html_body: String,
}

#[derive(Deserialize, Debug, Serialize, Clone)]
pub struct SmsConfig {
    pub message: String,
}

#[derive(Deserialize, Debug, Serialize, Clone)]
pub struct SendCommunicationBody {
    pub audience_selection: Option<AudienceSelection>,
    pub audience_voter_ids: Option<Vec<String>>,
    pub r#type: Option<CommunicationType>,
    pub communication_method: Option<CommunicationMethod>,
    pub schedule_now: Option<bool>,
    pub schedule_date: Option<String>,
    pub email: Option<EmailConfig>,
    pub sms: Option<SmsConfig>,
    pub document: Option<String>,
    pub name: Option<String>,
    pub alias: Option<String>,
}

File: packages/admin-portal/src/types/communications.ts
// SPDX-FileCopyrightText: 2024 Sequent Tech <legal@sequentech.io>
//
// SPDX-License-Identifier: AGPL-3.0-only

export enum IAudienceSelection {
    ALL_USERS = "ALL_USERS",
    NOT_VOTED = "NOT_VOTED",
    VOTED = "VOTED",
    SELECTED = "SELECTED",
}

export enum ICommunicationType {
    CREDENTIALS = "CREDENTIALS",
    BALLOT_RECEIPT = "BALLOT_RECEIPT",
    PARTICIPATION_REPORT = "PARTICIPATION_REPORT",
    ELECTORAL_RESULTS = "ELECTORAL_RESULTS",
    OTP = "OTP",
    TALLY_REPORT = "TALLY_REPORT",
}

export enum ICommunicationMethod {
    EMAIL = "EMAIL",
    SMS = "SMS",
    DOCUMENT = "DOCUMENT",
}

export interface IEmail {
    subject: string
    plaintext_body: string
    html_body: string
}

export interface IEmail {
    subject: string
    plaintext_body: string
    html_body: string
}

export interface ISmsConfig {
    message: string
}

export interface ISendCommunicationBody {
    audience_selection?: IAudienceSelection
    audience_voter_ids?: Array<string>
    type?: ICommunicationType
    communication_method?: ICommunicationMethod
    schedule_now?: boolean
    schedule_date?: string
    email?: IEmail
    sms?: ISmsConfig
    name?: string
    alias?: string
    document?: string
}

export interface IRECEIPTS {
    [key: string]: {
        allowed?: boolean
        template?: string | null
    }
}
